### PR TITLE
Port cub::DeviceReduce tests to catch2

### PR DESCRIPTION
**Pull request by elstehle**
## Description

This PR partially addresses https://github.com/NVIDIA/cccl/issues/85. It migrates tests for `DeviceReduce`, i.e., single-problem reduction algorithms, to Catch2. 

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/test/catch2_test_device_reduce.cu [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+    // Run test
+    thrust::device_vector<output_t> out_result(num_segments);
+    auto d_out_it = thrust::raw_pointer_cast(out_result.data());
+    device_reduce(in_it, d_out_it, num_items, reduction_op, init_t{});
+
+    // Verify result
+    REQUIRE(expected_result == out_result[0]);
+  }
+}
+CUB_TEST("Device reduce compiles with discard output iterator",
```
Comment (senior-zero) [inline cub/test/catch2_test_device_reduce.cu:430]:
```suggestion
}

CUB_TEST("Device reduce compiles with discard output iterator",
```

File: cub/test/catch2_test_device_reduce.cu [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+                                        min_items,
+                                        max_items,
+                                      }));
+
+  // Prepare input data
+  item_t default_constant{};
+  init_default_constant(default_constant);
+  auto in_it = thrust::make_constant_iterator(default_constant);
+
+  SECTION("reduce")
```
Comment (senior-zero) [inline cub/test/catch2_test_device_reduce.cu:408]: Is there any point in a single section per test that I'm missing?

File: cub/test/catch2_test_device_reduce.cu [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+  {
+    using op_t = cub::Sum;
+
+    // Binary reduction operator
+    auto reduction_op = unwrap_op(reference_extended_fp(d_in_it), op_t{});
+
+    // Prepare verification data
+    using accum_t            = cub::detail::accumulator_t<op_t, output_t, item_t>;
+    output_t expected_result = compute_single_problem_reference(in_items, reduction_op, accum_t{});
```
Comment (senior-zero) [inline cub/test/catch2_test_device_reduce.cu:224]:
```suggestion
    output_t expected_result = static_cast<output_t>(compute_single_problem_reference(in_items, reduction_op, accum_t{}));
```

File: cub/test/catch2_test_device_reduce.cu [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+// Skip DeviceReduce::Sum tests for extended floating-point types because of unbounded epsilon due
+// to pseudo associativity of the addition operation over floating point numbers
+#if TEST_TYPES != 3
+  SECTION("sum")
+  {
+    using op_t    = cub::Sum;
+    using accum_t = cub::detail::accumulator_t<op_t, output_t, item_t>;
+
+    // Prepare verification data
+    output_t expected_result = compute_single_problem_reference(in_items, op_t{}, accum_t{});
```
Comment (senior-zero) [inline cub/test/catch2_test_device_reduce.cu:245]:
```suggestion
    output_t expected_result = static_cast<output_t>(compute_single_problem_reference(in_items, op_t{}, accum_t{}));
```

File: cub/test/catch2_test_device_reduce.cu [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+    device_arg_max(d_in, d_out, n);
+
+    const out_t result = out[0];
+    REQUIRE(result.key == 0);
+    REQUIRE(result.value == -inf);
+  }
+}
+
+// TEST_TYPES == 5
+#elif TEST_TYPES == 5
```
Comment (senior-zero) [inline cub/test/catch2_test_device_reduce.cu:384]: This approach is not very development-friendly. If you need different tests in different TUs, can we consider putting them into different source files?

**Review by senior-zero — APPROVED** [review]
A few minor comments below.

File: cub/test/catch2_test_device_reduce.cu (hunk 5 above)
Comment (elstehle) [reply cub/test/catch2_test_device_reduce.cu]: Agreed! Thanks for the suggestion. Moved the two test sections into their own translation unit:
- `device_reduce_iteraturs.cu`: 64-bit indices and fancy iterators
- `device_reduce_fp_inf.cu`: `Arg{Min,Max}` with `inf`
Would be great if I could have your eyes do another pass over this refactor quickly 🙏

File: cub/test/catch2_test_device_reduce.cu (hunk 2 above)
Comment (elstehle) [reply cub/test/catch2_test_device_reduce.cu]: I agree, I don't think having a single section is justified here.

File: cub/test/catch2_test_device_reduce.cuh [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+  }
+
+  return a;
+}
+
+/**
+ * @brief Introduces the required NumericTraits for `c2h::custom_type_t`.
+ */
+template <template <typename> class... Policies>
+struct NumericTraits<c2h::custom_type_t<Policies...>>
```
[the remark this answers is not in the document]
Comment (elstehle) [reply cub/test/catch2_test_device_reduce.cuh:108]: Good catch, thanks!

**Review by senior-zero — APPROVED** [review]

**Pull request merged**